src/lib/auth/oauth-client.ts

```diff
@@ -0,0 +1,170 @@
+    };
+    return fetch(`${accountURL}/api/v1/oauth/token`, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+      body: new URLSearchParams(params),
+    });
+  }
+
+  private _generateCSRFToken(): string {
+    return Buffer.from(randomBytes(20)).toString("base64url");
+  }
+
+  private _handleRequest(req: IncomingMessage, res: ServerResponse) {
+    const allowedOrigins = [
+      "http://localhost:3005",
+      "http://127.0.0.1:3005",
+      "http://dashboard.fauna.com",
+      "http://dashboard.fauna-dev.com",
+      "http://dashboard.fauna-preview.com",
+    ];
+    const origin = req.headers.origin || "";
+
+    if (allowedOrigins.includes(origin)) {
+      res.setHeader("Access-Control-Allow-Origin", origin);
+      res.setHeader("Access-Control-Allow-Methods", "GET");
+      res.setHeader("Access-Control-Allow-Headers", "Content-Type");
+    }
+
+    let errorMessage = "";
+
+    if (req.method === "GET") {
+      const parsedUrl = url.parse(req.url || "", true);
+      if (parsedUrl.pathname !== "/") {
+        errorMessage = "Invalid redirect uri";
+        this.closeServer();
+      }
+      const query = parsedUrl.query;
+      if (query.error) {
+        errorMessage = query.error.toString();
+        this.closeServer();
+      }
+      if (query.code) {
+        const authCode = query.code;
+        if (!authCode || typeof authCode !== "string") {
+          errorMessage = "Invalid authorization code received";
+          this.server.close();
+        } else {
+          this.auth_code = authCode;
+          if (query.state !== this.state) {
+            errorMessage = "Invalid state received";
+            this.closeServer();
+          }
+          // TODO: Send them to a nice page that shows auth is complete and they can close the window.
+          res.writeHead(200, { "Content-Type": "text/html" });
+          res.end();
+          this.server.emit("auth_code_received");
+          this.closeServer();
+        }
+      }
+    } else {
+      errorMessage = "Invalid request method";
+      this.closeServer();
+    }
+    if (errorMessage) {
+      console.error("Error during authentication:", errorMessage);
+    }
+  }
+
+  private _isPortAvailable(port: number): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      const tester = net
+        .createServer()
+        .once("error", (err: any) =>
+          err.code === "EADDRINUSE" ? resolve(false) : reject(err)
+        )
+        .once("listening", () =>
+          tester.once("close", () => resolve(true)).close()
+        )
+        .listen(port);
+    });
+  }
+
+  public async start() {
+    let port: number;
+    let isAvailable: boolean = false;
+
+    // Loop until an available port is found
+    do {
+      port = Math.floor(Math.random() * (63000 - 62500 + 1)) + 62500;
+      isAvailable = await this._isPortAvailable(port);
+    } while (!isAvailable);
```

Comment:
thoughts on starting at a random point and walking linearly? this doesn't ever halt.

Reply:
can you bind to port zero? that should let the OS pick a random port afaik

Reply:
every time i tried on zero it just stuck with 0 🤷

Reply:
that worked! https://github.com/fauna/fauna-shell/pull/364/commits/16c262cc3e0390e533649b7230268b882b9fe020